// graphql-compose-mongoose/es/composeWithMongoose.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

/* eslint-disable no-use-before-define, no-param-reassign */

import { TypeComposer, InputTypeComposer } from 'graphql-compose';
import composeWithConnection from 'graphql-compose-connection';
import { convertModelToGraphQL } from './fieldsConverter';
import * as resolvers from './resolvers';
import { getUniqueIndexes, extendByReversedIndexes } from './utils/getIndexesFromModel';

export function composeWithMongoose(model) {
  var opts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  var name = opts && opts.name || model.modelName;

  var typeComposer = convertModelToGraphQL(model, name);

  if (opts.description) {
    typeComposer.setDescription(opts.description);
  }

  if (opts.fields) {
    prepareFields(typeComposer, opts.fields);
  }

  // $FlowFixMe
  typeComposer.setRecordIdFn(function (source) {
    return source ? '' + source._id : '';
  });

  createInputType(typeComposer, opts.inputType);

  if (!{}.hasOwnProperty.call(opts, 'resolvers') || opts.resolvers !== false) {
    createResolvers(model, typeComposer, opts.resolvers || {});
  }

  return typeComposer;
}

export function prepareFields(typeComposer, opts) {
  if (Array.isArray(opts.only)) {
    (function () {
      var onlyFieldNames = opts.only;
      var removeFields = Object.keys(typeComposer.getFields()).filter(function (fName) {
        return !onlyFieldNames.includes(fName);
      });
      typeComposer.removeField(removeFields);
    })();
  }
  if (opts.remove) {
    typeComposer.removeField(opts.remove);
  }
}

export function prepareInputFields(inputTypeComposer, inputFieldsOpts) {
  if (Array.isArray(inputFieldsOpts.only)) {
    (function () {
      var onlyFieldNames = inputFieldsOpts.only;
      var removeFields = Object.keys(inputTypeComposer.getFields()).filter(function (fName) {
        return !onlyFieldNames.includes(fName);
      });
      inputTypeComposer.removeField(removeFields);
    })();
  }
  if (inputFieldsOpts.remove) {
    inputTypeComposer.removeField(inputFieldsOpts.remove);
  }
  if (inputFieldsOpts.required) {
    inputTypeComposer.makeRequired(inputFieldsOpts.required);
  }
}

export function createInputType(typeComposer) {
  var inputTypeOpts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  var inputTypeComposer = typeComposer.getInputTypeComposer();

  if (inputTypeOpts.name) {
    inputTypeComposer.setTypeName(inputTypeOpts.name);
  }

  if (inputTypeOpts.description) {
    inputTypeComposer.setDescription(inputTypeOpts.description);
  }

  if (inputTypeOpts.fields) {
    prepareInputFields(inputTypeComposer, inputTypeOpts.fields);
  }
}

export function createResolvers(model, typeComposer, opts) {
  var names = resolvers.getAvailableNames();
  names.forEach(function (resolverName) {
    if (!{}.hasOwnProperty.call(opts, resolverName) || opts[resolverName] !== false) {
      var createResolverFn = resolvers[resolverName];
      if (createResolverFn) {
        var resolver = createResolverFn(model, typeComposer, opts[resolverName] || {});
        typeComposer.setResolver(resolverName, resolver);
      }
    }
  });

  if (!{}.hasOwnProperty.call(opts, 'connection') || opts.connection !== false) {
    prepareConnectionResolver(model, typeComposer, opts.connection ? opts.connection : {});
  }
}

export function prepareConnectionResolver(model, typeComposer, opts) {
  var uniqueIndexes = extendByReversedIndexes(getUniqueIndexes(model), { reversedFirst: true });
  var sortConfigs = {};
  uniqueIndexes.forEach(function (indexData) {
    var keys = Object.keys(indexData);
    var name = keys.join('__').toUpperCase().replace(/[^_a-zA-Z0-9]/i, '__');
    if (indexData[keys[0]] === 1) {
      name = name + '_ASC';
    } else if (indexData[keys[0]] === -1) {
      name = name + '_DESC';
    }
    sortConfigs[name] = {
      value: indexData,
      cursorFields: keys,
      beforeCursorQuery: function beforeCursorQuery(rawQuery, cursorData) {
        keys.forEach(function (k) {
          if (!rawQuery[k]) rawQuery[k] = {};
          if (indexData[k] === 1) {
            rawQuery[k].$lt = cursorData[k];
          } else {
            rawQuery[k].$gt = cursorData[k];
          }
        });
      },
      afterCursorQuery: function afterCursorQuery(rawQuery, cursorData) {
        keys.forEach(function (k) {
          if (!rawQuery[k]) rawQuery[k] = {};
          if (indexData[k] === 1) {
            rawQuery[k].$gt = cursorData[k];
          } else {
            rawQuery[k].$lt = cursorData[k];
          }
        });
      }
    };
  });

  composeWithConnection(typeComposer, {
    findResolverName: 'findMany',
    countResolverName: 'count',
    sort: _extends({}, sortConfigs, opts)
  });
}
