// graphql-compose-mongoose/es/fieldsConverter.js

/* eslint-disable no-use-before-define */

import mongoose from 'mongoose';
import objectPath from 'object-path';
import { TypeComposer, GraphQLDate, GraphQLBuffer, GraphQLGeneric, GraphQLJSON } from 'graphql-compose';

import { GraphQLString, GraphQLFloat, GraphQLBoolean, GraphQLList, GraphQLEnumType, GraphQLObjectType } from 'graphql';
import GraphQLMongoID from './types/mongoid';
import typeStorage from './typeStorage';

export var ComplexTypes = {
  ARRAY: 'ARRAY',
  EMBEDDED: 'EMBEDDED',
  DOCUMENT_ARRAY: 'DOCUMENT_ARRAY',
  ENUM: 'ENUM',
  REFERENCE: 'REFERENCE',
  SCALAR: 'SCALAR',
  MIXED: 'MIXED'
};

function _getFieldName(field) {
  return field.path;
}

function _getFieldType(field) {
  return field.instance;
}

function _getFieldDescription(field) {
  if (field.options && field.options.description) {
    return field.options.description;
  }

  return undefined;
}

function _getFieldEnums(field) {
  if (field.enumValues && field.enumValues.length > 0) {
    return field.enumValues;
  }

  return undefined;
}

function capitalize(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

export function dotPathsToEmbedded(fields) {
  // convert only one dot-level on this step to EmbeddedModel
  // further when converting EmbeddedModel to GQL, it internally
  var result = {};

  Object.keys(fields).forEach(function (fieldName) {
    var dotIdx = fieldName.indexOf('.');
    if (dotIdx === -1) {
      result[fieldName] = fields[fieldName];
    } else {
      // create pseudo sub-model
      var name = fieldName.substr(0, dotIdx);
      if (!result[name]) {
        var embeddedField = {
          instance: 'Embedded',
          path: name,
          schema: {
            paths: {}
          }
        };
        result[name] = embeddedField;
      }
      var subName = fieldName.substr(dotIdx + 1);
      result[name].schema.paths[subName] = Object.assign({}, fields[fieldName], { path: subName });
    }
  });

  return result;
}

export function getFieldsFromModel(model) {
  if (!model || !model.schema || !model.schema.paths) {
    throw new Error('You provide incorrect mongoose model to `getFieldsFromModel()`. ' + 'Correct model should contain `schema.paths` properties.');
  }

  var fields = {};
  var paths = dotPathsToEmbedded(model.schema.paths);

  Object.keys(paths).filter(function (path) {
    return !path.startsWith('__');
  }) // skip hidden fields
  .forEach(function (path) {
    fields[path] = paths[path];
  });

  return fields;
}

export function convertModelToGraphQL(model, typeName) {
  if (!typeName) {
    throw new Error('You provide empty name for type. ' + '`name` argument should be non-empty string.');
  }

  var typeComposer = new TypeComposer(typeStorage.getOrSet(typeName, new GraphQLObjectType({
    name: typeName,
    interfaces: [],
    description: undefined,
    fields: {}
  })));

  var mongooseFields = getFieldsFromModel(model, typeName);
  var graphqlFields = {};

  Object.keys(mongooseFields).forEach(function (fieldName) {
    var mongooseField = mongooseFields[fieldName];
    graphqlFields[fieldName] = {
      type: convertFieldToGraphQL(mongooseField, typeName),
      description: _getFieldDescription(mongooseField)
    };
  });

  typeComposer.addFields(graphqlFields);
  return typeComposer;
}

export function convertSchemaToGraphQL(schema, typeName) {
  if (!typeName) {
    throw new Error('You provide empty name for type. ' + '`name` argument should be non-empty string.');
  }

  if (schema._gqcTypeComposer) {
    return schema._gqcTypeComposer;
  }

  var tc = convertModelToGraphQL({ schema: schema }, typeName);
  // also generate InputType
  tc.getInputTypeComposer();

  schema._gqcTypeComposer = tc; // eslint-disable-line
  return tc;
}

export function convertFieldToGraphQL(field) {
  var prefix = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

  var complexType = deriveComplexType(field);
  switch (complexType) {
    case ComplexTypes.SCALAR:
      return scalarToGraphQL(field);
    case ComplexTypes.ARRAY:
      return arrayToGraphQL(field, prefix);
    case ComplexTypes.EMBEDDED:
      return embeddedToGraphQL(field, prefix);
    case ComplexTypes.ENUM:
      return enumToGraphQL(field, prefix);
    case ComplexTypes.REFERENCE:
      return referenceToGraphQL(field, prefix);
    case ComplexTypes.DOCUMENT_ARRAY:
      return documentArrayToGraphQL(field, prefix);
    case ComplexTypes.MIXED:
      return mixedToGraphQL(field);
    default:
      return scalarToGraphQL(field);
  }
}

export function deriveComplexType(field) {
  if (!field || !field.path || !field.instance) {
    throw new Error('You provide incorrect mongoose field to `deriveComplexType()`. ' + 'Correct field should contain `path` and `instance` properties.');
  }

  var fieldType = _getFieldType(field);

  if (field instanceof mongoose.Schema.Types.DocumentArray) {
    return ComplexTypes.DOCUMENT_ARRAY;
  } else if (field instanceof mongoose.Schema.Types.Embedded || fieldType === 'Embedded') {
    return ComplexTypes.EMBEDDED;
  } else if (field instanceof mongoose.Schema.Types.Array || objectPath.has(field, 'caster.instance')) {
    return ComplexTypes.ARRAY;
  } else if (field instanceof mongoose.Schema.Types.Mixed) {
    return ComplexTypes.MIXED;
  } else if (fieldType === 'ObjectID') {
    return ComplexTypes.REFERENCE;
  }

  var enums = _getFieldEnums(field);
  if (enums) {
    return ComplexTypes.ENUM;
  }

  return ComplexTypes.SCALAR;
}

function removePseudoIdField(typeComposer) {
  // remove pseudo object id mongoose field
  var gqFields = typeComposer.getFields();
  var pseudoFieldNames = ['_id'];
  pseudoFieldNames.forEach(function (name) {
    if (gqFields[name] && gqFields[name].type === GraphQLMongoID) {
      typeComposer.removeField(name);
    }
  });
}

export function scalarToGraphQL(field) {
  var typeName = _getFieldType(field);

  switch (typeName) {
    case 'String':
      return GraphQLString;
    case 'Number':
      return GraphQLFloat;
    case 'Date':
      return GraphQLDate;
    case 'Buffer':
      return GraphQLBuffer;
    case 'Boolean':
      return GraphQLBoolean;
    case 'ObjectID':
      return GraphQLMongoID;
    default:
      return GraphQLGeneric;
  }
}

export function arrayToGraphQL(field) {
  var prefix = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

  if (!field || !field.caster) {
    throw new Error('You provide incorrect mongoose field to `arrayToGraphQL()`. ' + 'Correct field should contain `caster` property.');
  }

  var unwrappedField = Object.assign({}, field.caster);
  objectPath.set(unwrappedField, 'options.ref', objectPath.get(field, 'options.ref', undefined));

  var outputType = convertFieldToGraphQL(unwrappedField, prefix);
  return new GraphQLList(outputType);
}

export function embeddedToGraphQL(field) {
  var prefix = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

  var fieldName = _getFieldName(field);
  var fieldType = _getFieldType(field);

  if (fieldType !== 'Embedded') {
    throw new Error('You provide incorrect field to `embeddedToGraphQL()`. ' + 'This field should has `Embedded` type. ');
  }

  var typeName = '' + prefix + capitalize(fieldName);
  var typeComposer = convertSchemaToGraphQL(field.schema, typeName);
  removePseudoIdField(typeComposer);

  return typeComposer.getType();
}

export function enumToGraphQL(field) {
  var prefix = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

  var valueList = _getFieldEnums(field);
  if (!valueList) {
    throw new Error('You provide incorrect mongoose field to `enumToGraphQL()`. ' + 'Correct field should contain `enumValues` property');
  }

  var graphQLEnumValues = valueList.reduce(function (result, val) {
    result[val] = { value: val }; // eslint-disable-line no-param-reassign
    return result;
  }, {});

  var typeName = 'Enum' + prefix + capitalize(_getFieldName(field));
  return typeStorage.getOrSet(typeName, new GraphQLEnumType({
    name: typeName,
    description: _getFieldDescription(field),
    values: graphQLEnumValues
  }));
}

export function documentArrayToGraphQL(field) {
  var prefix = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

  if (!(field instanceof mongoose.Schema.Types.DocumentArray)) {
    throw new Error('You provide incorrect mongoose field to `documentArrayToGraphQL()`. ' + 'Correct field should be instance of `mongoose.Schema.Types.DocumentArray`');
  }

  var typeName = '' + prefix + capitalize(_getFieldName(field));

  var typeComposer = convertModelToGraphQL(field, typeName);
  removePseudoIdField(typeComposer);

  return new GraphQLList(typeComposer.getType());
}

export function referenceToGraphQL(field) {
  var prefix = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

  var fieldType = _getFieldType(field);
  if (fieldType !== 'ObjectID') {
    throw new Error('You provide incorrect mongoose field to `referenceToGraphQL()`. ' + 'Correct field should has mongoose-type `ObjectID`');
  }

  // const refModelName = objectPath.get(field, 'options.ref');
  // if (refModelName) {
  //   return GQLReference;
  //   // throw new Error('Mongoose REFERENCE to graphQL TYPE not implemented yet. '
  //   //                + `Field ${_getFieldName(field)}`);
  //   // Storage.UnresolvedRefs.setSubKey(parentTypeName, fieldName, { refModelName });
  //   // return GraphQLReference;
  // }

  // this is mongo id field
  return scalarToGraphQL(field, prefix);
}

export function mixedToGraphQL(field) {
  if (!(field instanceof mongoose.Schema.Types.Mixed)) {
    throw new Error('You provide incorrect mongoose field to `mixedToGraphQL()`. ' + 'Correct field should be instance of `mongoose.Schema.Types.Mixed`');
  }

  return GraphQLJSON;
}
