// graphql-compose/es/utils/misc.js
import pluralize from './pluralize';

export function resolveMaybeThunk(thingOrThunk) {
  // eslint-disable-line
  return typeof thingOrThunk === 'function' ? thingOrThunk() : thingOrThunk;
}

export function camelCase(str) {
  return str.replace(/(?:^\w|[A-Z]|\b\w)/g, function (letter, index) {
    return index === 0 ? letter.toLowerCase() : letter.toUpperCase();
  }).replace(/\s+/g, '');
}

export function getPluralName(name) {
  return pluralize(camelCase(name));
}

export function upperFirst(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

export function clearName(str) {
  return str.replace(/[^_a-zA-Z0-9]/g, '');
}

export function omit(obj, keys) {
  if (!obj) {
    return {};
  }

  var result = Object.assign({}, obj);
  if (Array.isArray(keys)) {
    keys.forEach(function (k) {
      delete result[k];
    });
  } else {
    delete result[keys];
  }

  return result;
}

export function only(obj, keys) {
  if (!obj) {
    return {};
  }

  var result = {};
  if (Array.isArray(keys)) {
    keys.forEach(function (k) {
      if ({}.hasOwnProperty.call(obj, k)) {
        result[k] = obj[k];
      }
    });
  } else if ({}.hasOwnProperty.call(obj, keys)) {
    result[keys] = obj[keys];
  }

  return result;
}
