// graphql-compose-mongoose/es/resolvers/updateOne.js
var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

/* eslint-disable no-param-reassign */

import { GraphQLObjectType } from 'graphql';
import { Resolver, TypeComposer } from 'graphql-compose';
import { skipHelperArgs } from './helpers/skip';
import { recordHelperArgs } from './helpers/record';
import { filterHelperArgs } from './helpers/filter';
import { sortHelperArgs } from './helpers/sort';
import findOne from './findOne';
import GraphQLMongoID from '../types/mongoid';
import typeStorage from '../typeStorage';

export default function updateOne(model, typeComposer, opts) {
  if (!model || !model.modelName || !model.schema) {
    throw new Error('First arg for Resolver updateOne() should be instance of Mongoose Model.');
  }
  if (!(typeComposer instanceof TypeComposer)) {
    throw new Error('Second arg for Resolver updateOne() should be instance of TypeComposer.');
  }

  var findOneResolver = findOne(model, typeComposer, opts);

  var outputTypeName = 'UpdateOne' + typeComposer.getTypeName() + 'Payload';
  var outputType = typeStorage.getOrSet(outputTypeName, new GraphQLObjectType({
    name: outputTypeName,
    fields: {
      recordId: {
        type: GraphQLMongoID,
        description: 'Updated document ID'
      },
      record: {
        type: typeComposer.getType(),
        description: 'Updated document'
      }
    }
  }));

  var resolver = new Resolver({
    name: 'updateOne',
    kind: 'mutation',
    description: 'Update one document: ' + '1) Retrieve one document via findOne. ' + '2) Apply updates to mongoose document. ' + '3) Mongoose applies defaults, setters, hooks and validation. ' + '4) And save it.',
    outputType: outputType,
    args: _extends({}, recordHelperArgs(typeComposer, _extends({
      recordTypeName: 'UpdateOne' + typeComposer.getTypeName() + 'Input',
      removeFields: ['id', '_id'],
      isRequired: true
    }, opts && opts.record)), filterHelperArgs(typeComposer, model, _extends({
      filterTypeName: 'FilterUpdateOne' + typeComposer.getTypeName() + 'Input',
      model: model
    }, opts && opts.filter)), sortHelperArgs(model, _extends({
      sortTypeName: 'SortUpdateOne' + typeComposer.getTypeName() + 'Input'
    }, opts && opts.sort)), skipHelperArgs()),
    resolve: function resolve(resolveParams) {
      var recordData = resolveParams.args && resolveParams.args.record || null;
      var filterData = resolveParams.args && resolveParams.args.filter || {};

      if (!((typeof filterData === 'undefined' ? 'undefined' : _typeof(filterData)) === 'object') || Object.keys(filterData).length === 0) {
        return Promise.reject(new Error(typeComposer.getTypeName() + '.updateOne resolver requires ' + 'at least one value in args.filter'));
      }

      resolveParams.projection = resolveParams.projection && resolveParams.projection.record || {};

      return findOneResolver.resolve(resolveParams).then(function (doc) {
        if (resolveParams.beforeRecordMutate) {
          return resolveParams.beforeRecordMutate(doc);
        }
        return doc;
      })
      // save changes to DB
      .then(function (doc) {
        if (recordData) {
          doc.set(recordData);
          return doc.save();
        }
        return doc;
      })
      // prepare output payload
      .then(function (record) {
        if (record) {
          return {
            record: record,
            recordId: typeComposer.getRecordIdFn()(record)
          };
        }

        return null;
      });
    }
  });

  return resolver;
}
